qa/rpc-tests/turnstile.py

```diff
@@ -0,0 +1,202 @@
+        connect_nodes_bi(self.nodes,1,2)
+        self.is_network_split=False
+        self.sync_all()
+
+    # Helper method to verify the size of a shielded value pool for a given node
+    def assert_pool_balance(self, node, name, balance):
+        pools = node.getblockchaininfo()['valuePools']
+        for pool in pools:
+            if pool['id'] == name:
+                assert_equal(pool['chainValue'], balance)
+                return
+        assert(False)
+
+    # Helper method to start a single node with extra args and sync to the network
+    def start_and_sync_node(self, index, args=[]):
+        self.nodes[index] = start_node(index, self.options.tmpdir, extra_args=NUPARAMS_ARGS + args)
+        connect_nodes_bi(self.nodes,0,1)
+        connect_nodes_bi(self.nodes,1,2)
+        connect_nodes_bi(self.nodes,0,2)
+        self.sync_all()
+
+    # Helper method to stop and restart a single node with extra args and sync to the network
+    def restart_and_sync_node(self, index, args=[]):
+        self.nodes[index].stop()
+        bitcoind_processes[index].wait()
+        self.start_and_sync_node(index, args)
+
+    def run_test(self):
+        # Sanity-check the test harness
+        self.nodes[0].generate(200)
+        assert_equal(self.nodes[0].getblockcount(), 200)
+        self.sync_all()
+
+        # Activate Sapling
+        self.nodes[0].generate(1)
+        self.sync_all()
+
+        # Node 0 shields some funds
+        dest_addr = self.nodes[0].z_getnewaddress(POOL_NAME.lower())
+        taddr0 = get_coinbase_address(self.nodes[0])
+        recipients = []
+        recipients.append({"address": dest_addr, "amount": Decimal('10')})
+        myopid = self.nodes[0].z_sendmany(taddr0, recipients, 1, 0)
+        wait_and_assert_operationid_status(self.nodes[0], myopid)
+        self.sync_all()
+        self.nodes[0].generate(1)
+        self.sync_all()
+        assert_equal(self.nodes[0].z_getbalance(dest_addr), Decimal('10'))
+
+        # Verify size of shielded pool
+        self.assert_pool_balance(self.nodes[0], POOL_NAME.lower(), Decimal('10'))
+        self.assert_pool_balance(self.nodes[1], POOL_NAME.lower(), Decimal('10'))
+        self.assert_pool_balance(self.nodes[2], POOL_NAME.lower(), Decimal('10'))
+
+        # Relaunch node 0 with in-memory size of value pools set to zero.
+        self.restart_and_sync_node(0, TURNSTILE_ARGS)
+
+        # Verify size of shielded pool
+        self.assert_pool_balance(self.nodes[0], POOL_NAME.lower(), Decimal('0'))
+        self.assert_pool_balance(self.nodes[1], POOL_NAME.lower(), Decimal('10'))
+        self.assert_pool_balance(self.nodes[2], POOL_NAME.lower(), Decimal('10'))
+
+        # Node 0 creates an unshielding transaction
+        recipients = []
+        recipients.append({"address": taddr0, "amount": Decimal('1')})
+        myopid = self.nodes[0].z_sendmany(dest_addr, recipients, 1, 0)
+        mytxid = wait_and_assert_operationid_status(self.nodes[0], myopid)
```

Comment:
Optional improvement: test that the transaction appears in the memory pool(s) of all the nodes.

Reply:
Done